clients/src/main/java/org/apache/kafka/common/requests/OffsetCommitResponse.java

```diff
@@ -119,28 +131,52 @@ public boolean shouldClientThrottle(short version) {
         return version >= 4;
     }
 
+    public short version() {
+        return version;
+    }
+
     public static class Builder {
         OffsetCommitResponseData data = new OffsetCommitResponseData();
         HashMap<String, OffsetCommitResponseTopic> byTopicName = new HashMap<>();
+        private final TopicResolver topicResolver;
+        private final short version;
+
+        public Builder(TopicResolver topicResolver, short version) {
```

Comment:
I see what you mean. I moved this logic in the callback of the future which merges the results from the coordinator with those created by the request handling method. I thought about extending the support id in internal layers (group coordinator) in a PR of its own. So, eventually, the coordinator will return results populated with topic ids when applicable.

Reply:
I actually wonder if we should do it the other way around. We could do KAFKA-14793 first, merge it, and update this one accordingly. Without KAFKA-14793, the contract of the not really respected and it feels a bit weird to work around it here instead of fixing the real issue. Is KAFKA-14793 complicated? What do you think?

Reply:
Thanks for the answer. If I understand correctly, we would then have a resolution of topic ids from topic-name-based data, so this may not prevent offsets from a topic to be provided as those of another topic with the same name (defined at different point in time in the server)?

The resolution can be done in the group coordinator layer, assuming it has access to the topic id resolved upstream by the request handler. Because we want to preserve the same mapping used when request processing started, we need to ensure the right ids are used within the adaptor's `GroupCoordinator#commitOffsets` method(). Since the mapping returned from the metadata cache depends on the snapshot used at the time the mapping is requested, if the adaptor retrieves it from the metadata cache internally, there is no guarantee the metadata is the same hence that the topic IDs registered with the broker are the same.

This means that the topic ids need to be propagated from the request handler (`KafkaApis`) to the coordinator adaptor somehow. Without a change in the method and contract implemented by the coordinator, these ids could be transferred via the `OffsetCommitRequestData` DTO directly, which means a change in the API schema would be required prior to the change. Alternatively, we may want to change the interface of the coordinator and change the signature of the offset commit method to allow for the propagation of topic ids.

I may be missing the entire thing though?

Reply:
How about doing the following?

We change the signature of `GroupCoordinator.handleCommitOffsets` to the following:

```
  def handleCommitOffsets(groupId: String,
                          memberId: String,
                          groupInstanceId: Option[String],
                          generationId: Int,
                          offsetMetadata: immutable.Map[TopicIdPartition, OffsetAndMetadata],
                          responseCallback: immutable.Map[TopicIdPartition, Errors] => Unit,
                          requestLocal: RequestLocal = RequestLocal.NoCaching): Unit = {
```

Note the change from `TopicPartition` to `TopicIdPartition` for `offsetMetadata` and `responseCallback`.

Then, we have to adapt the implementation of `handleCommitOffsets` to get the `TopicPartition` from the `TopicIdPartition` where required. We can keep `pendingOffsetCommits` and `offsets` keyed by `TopicPartition` for now in `GroupMetadataManager`.

This allows the preservation of the topic ids provided to the GroupCoordinator but it does not provide any stronger guarantee for the offsets yet (as you pointed out). With this approach, we don't depend on the resolver at all.

Reply:
Sounds good. Thanks for your guidance. As you mentioned, this PR is already quite large, so if you agree, I will go ahead and implement this change first, in a PR of its own. Thanks!

Reply:
Sounds good to me. Thanks!